### PR TITLE
Export `run_as_background_process` from the module API

### PR DESCRIPTION
**Pull request by H-Shay**
This PR exports `run_as_background_process` from the `module_api` class so that it can be used as part of the solution to https://github.com/matrix-org/synapse-auto-accept-invite/issues/12. The associated PR for that issue is: https://github.com/matrix-org/synapse-auto-accept-invite/pull/15.

### CODE REVIEW

File: synapse/module_api/__init__.py [hunk 1]
```diff
@@ -1666,6 +1667,46 @@ async def set_displayname(
...
+    ) -> "defer.Deferred[Optional[R]]":
+        """Run the given function in its own logcontext, with resource metrics
+
+        This should be used to wrap processes which are fired off to run in the
+        background, instead of being associated with a particular request.
+
+        It returns a Deferred which completes when the function completes, but it doesn't
+        follow the synapse logcontext rules, which makes it appropriate for passing to
+        clock.looping_call and friends (or for firing-and-forgetting in the middle of a
+        normal synapse async function).
```
Comment (MatMaul) [inline synapse/module_api/__init__.py:1688]: This description seems quite entangled to Synapse internals, here is a suggestion trying to be a bit more generic.
```suggestion
        """Run the given function in the background, with its own logcontext.

        This should be used to wrap processes which are fired off to run in the
        background.

        It returns a Deferred which completes when the function completes, but it doesn't
        follow the synapse logcontext rules.
```

**Review by MatMaul — CHANGES_REQUESTED** [review]

File: synapse/module_api/__init__.py [hunk 2]
```diff
@@ -1666,6 +1667,46 @@ async def set_displayname(
             deactivation=deactivation,
         )
 
+    R = TypeVar("R")
+
+    def run_as_background_process(
```
Comment (squahtx) [inline synapse/module_api/__init__.py:1672]: It looks like we expose `run_in_background` in a different way that avoids all this. Can we follow the same pattern?

https://github.com/matrix-org/synapse/blob/2e59e97ebd02e93da39e6c90335d3b24ed01217a/synapse/module_api/__init__.py#L158

Comment (H-Shay) [reply synapse/module_api/__init__.py]: Oh that is easier, thanks for the suggestion!

File: synapse/module_api/__init__.py (hunk 1 above)
Comment (H-Shay) [reply synapse/module_api/__init__.py]: I've went with Sean's suggestion to below which eliminates the need for any of this, but thank you for pointing it out! It's helpful to know how to think about these things.

**Review by MatMaul — APPROVED** [review]

File: changelog.d/15577.misc [hunk 3]
```diff
@@ -0,0 +1 @@
+Export `run_as_background_process` via the `module_api` class.
```
Comment (squahtx) [inline changelog.d/15577.misc:1]:
```suggestion
Export `run_as_background_process` from the module API.
```

**Review by squahtx — APPROVED** [review]

**Pull request merged**